custom/util/panel.libsonnet

```diff
@@ -318,4 +318,103 @@ local xtd = import 'github.com/jsonnet-libs/xtd/main.libsonnet';
         else p,
       panels
     ),
+
+
+  '#setRefIDs':: d.func.new(
+    |||
+      `setRefIDs` calculates the `refId` field for each target on a panel.
+    |||,
+    args=[
+      d.arg('panel', d.T.object),
+      d.arg('overrideExistingIDs', d.T.bool, default=true),
+    ]
+  ),
+  setRefIDs(panel, overrideExistingIDs=true):
+    local calculateRefID(n) =
+      // From: https://github.com/grafana/grafana/blob/bffd87107b786930edd091060143ee013843efac/packages/grafana-data/src/query/refId.ts#L15
+      local letters = std.map(std.char, std.range(std.codepoint('A'), std.codepoint('Z')));
+      if n < std.length(letters)
+      then letters[n]
+      else calculateRefID(std.floor(n / std.length(letters)) - 1) + letters[std.mod(n, std.length(letters))];
+    panel + {
+      targets:
+        std.mapWithIndex(
+          function(i, target)
+            if overrideExistingIDs
+               || !std.objectHas(target, 'refId')
+            then target + {
+              refId: calculateRefID(i),
+            }
+            else target,
+          panel.targets,
+        ),
+    },
+
+  '#setRefIDsOnPanels':: d.func.new(
+    |||
+      `setRefIDsOnPanels` applies `setRefIDs on all `panels`.
+    |||,
+    args=[
+      d.arg('panels', d.T.array),
+    ]
+  ),
+  setRefIDsOnPanels(panels):
```

Comment:
Should we keep the index, having multiple `refID` A could get confusing if we're referencing between panels

Reply:
I'm not sure if I understand what you mean.

`refID` is unique per panel, so a reference to a target is a combination of the panel ID and the refID of the target